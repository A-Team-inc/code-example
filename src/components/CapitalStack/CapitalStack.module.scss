@import 'src/styles/global/variables';

.CapitalStack {
  border: $border-default;
  border-radius: 4px;
  padding: $spacing-l 0 0;
  // Todo: Remove fixed width when the second column block will be implemented
  width: 565px;

  @media screen and (max-width: $max-mobile-w) {
    width: auto;
  }

  &__row {
    display: flex;
    padding: 0 $spacing-l $spacing-l;

    @media screen and (max-width: $max-mobile-w) {
      flex-direction: column;
      align-items: center;
    }
  }

  &__content {
    display: flex;
    flex-direction: column;
    padding-left: 33px;
    width: 100%;

    @media screen and (max-width: $max-mobile-w) {
      padding: $spacing-l 0 0;
    }
  }

  &__graph {
    display: flex;
    align-items: center;

    &::before {
      content: '';
      background: url('/svg/priority.svg');
      height: 168px;
      width: 19px;
    }

    @media screen and (max-width: $max-mobile-w) {
      // icon width + $spacing-l
      margin: 0 0 0 -35px;
    }
  }

  &__graphContent {
    width: 200px;
    height: 200px;
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    margin: 0 0 0 $spacing-l;
    transform: rotate(-90deg);
  }

  &__contentItem {
    background-color: $alabaster;
    padding: $spacing-xs $spacing-l;
    border-radius: 4px;
    margin-bottom: 7px;
    flex: 1 0 auto;

    &:last-child {
      margin-bottom: 0;
    }
  }

  &__titleWrapper {
    font-family: $default-medium;
    font-size: $fontSize-sm;
    color: $sub-text;
    line-height: normal;
    display: flex;
    align-items: center;
  }

  &__label {
    height: 16px;
    width: 16px;
    display: block;
    border-radius: 3px;
    margin-right: $spacing-xs;
  }

  &__amountContent {
    font-family: $default-semiBold;
    font-size: $default_fontSize;
    color: $text-black;
    line-height: normal;
    display: flex;
    justify-content: space-between;
  }

  &__amount {
    text-align: right;
    color: $sub-text;
  }

  &__footer {
    background-color: $alabaster;
    text-align: right;
    padding: $spacing-m $spacing-l;
    border-top: $border-default;
  }

  &__footerContent {
    font-family: $default-semiBold;
    font-size: $fontSize-sm;
    color: $text-black;

    span {
      color: $sub-text;

      &:first-child {
        margin-right: $spacing-xs;
      }

      &:last-child {
        margin-left: $spacing-xxs;
      }
    }
  }
}
