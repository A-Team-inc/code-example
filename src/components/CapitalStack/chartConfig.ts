import { Options } from 'highcharts';

export const chartConfig: Options = {
  chart: {
    height: 200,
    spacingTop: 0,
    spacingBottom: 0,
    margin: 0
  },
  title: {
    text: ''
  },
  tooltip: {
    enabled: false
  },
  lang: {
    noData: 'No data available'
  },
  noData: {
    style: {
      fontSize: '14px',
      color: '#999999'
    },
    position: {
      verticalAlign: 'middle',
      align: 'center'
    }
  }
};

interface ICapitalData {
  title: string;
  color: string;
  percentage: number;
  amount: string;
}

export const setChartData = (data: ICapitalData[]) => {
  const buildIDs = data.map(({ title, color }) => ({ id: title, color }));
  const buildChild = data.map(({ title, percentage }) => ({ parent: title, value: percentage }));

  return {
    series: [
      {
        type: 'treemap',
        levels: [
          {
            level: 1,
            layoutAlgorithm: 'sliceAndDice',
            dataLabels: {
              enabled: true,
              align: 'left',
              verticalAlign: 'top',
              style: {
                fontSize: '15px',
                fontWeight: 'bold'
              }
            }
          }
        ],
        data: [...buildIDs, ...buildChild]
      }
    ]
  };
};
