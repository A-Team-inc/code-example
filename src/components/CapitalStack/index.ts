import CapitalStack from './CapitalStack';
export default CapitalStack;
