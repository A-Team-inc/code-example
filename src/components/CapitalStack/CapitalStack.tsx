import { Options } from 'highcharts';
import Chart from 'molecules/Chart';
import React from 'react';
import { currencyFormat, currencyManager } from 'utils/currencyManipulation';
import { ICapitalStack } from 'utils/globalTypes';

import s from './CapitalStack.module.scss';
import { chartConfig, setChartData } from './chartConfig';

const CapitalStack = ({ stack, total }: ICapitalStack) => {
  const content = currencyManager(stack, total) as {
    title: string;
    color: string;
    percentage: number;
    amount: string;
  }[];

  const preparedData = setChartData(content) as Options;

  return (
    <div className={s.CapitalStack}>
      <div className={s.CapitalStack__row}>
        <div className={s.CapitalStack__graph}>
          <div className={s.CapitalStack__graphContent}>
            <Chart config={chartConfig} dataValues={preparedData} chartType="treemap" />
          </div>
        </div>
        <div className={s.CapitalStack__content}>
          {content &&
            content.map(({ title, color, percentage, amount }) => (
              <div key={`content-${title}`} className={s.CapitalStack__contentItem}>
                <div className={s.CapitalStack__titleWrapper}>
                  <span className={s.CapitalStack__label} style={{ backgroundColor: color }} />
                  {title}
                </div>
                <div className={s.CapitalStack__amountContent}>
                  {`${percentage}%`}
                  <span className={s.CapitalStack__amount}>{`${amount}`}</span>
                </div>
              </div>
            ))}
        </div>
      </div>
      <div className={s.CapitalStack__footer}>
        {total && (
          <div className={s.CapitalStack__footerContent}>
            <span>Total:</span>
            {currencyFormat(total, 'en-US')}
            <span>(100%)</span>
          </div>
        )}
      </div>
    </div>
  );
};

export default CapitalStack;
