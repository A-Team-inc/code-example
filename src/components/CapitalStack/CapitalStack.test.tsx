import { shallow } from 'enzyme';
import preloadAll from 'jest-next-dynamic';
import React from 'react';

import CapitalStack from './CapitalStack';

const props = {
  stack: [
    {
      title: 'Investor Equity',
      color: '#43c3d3',
      percentage: 31.9
    },
    {
      title: 'Sponsor Co-Investment',
      color: '#8ac13f',
      percentage: 3.5
    },
    {
      title: 'Senior Debt',
      color: '#1ea0c0',
      percentage: 53
    }
  ],
  total: 10000
};

describe('Capital Stack Component', () => {
  it('should match snapshots', async () => {
    await preloadAll();
    const wrapper = shallow(<CapitalStack {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
