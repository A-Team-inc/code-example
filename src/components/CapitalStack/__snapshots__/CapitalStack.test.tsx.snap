// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capital Stack Component should match snapshots 1`] = `
<div
  className="CapitalStack"
>
  <div
    className="CapitalStack__row"
  >
    <div
      className="CapitalStack__graph"
    >
      <div
        className="CapitalStack__graphContent"
      >
        <Chart
          chartType="treemap"
          config={
            Object {
              "chart": Object {
                "height": 200,
                "margin": 0,
                "spacingBottom": 0,
                "spacingTop": 0,
              },
              "lang": Object {
                "noData": "No data available",
              },
              "noData": Object {
                "position": Object {
                  "align": "center",
                  "verticalAlign": "middle",
                },
                "style": Object {
                  "color": "#999999",
                  "fontSize": "14px",
                },
              },
              "title": Object {
                "text": "",
              },
              "tooltip": Object {
                "enabled": false,
              },
            }
          }
          dataValues={
            Object {
              "series": Array [
                Object {
                  "data": Array [
                    Object {
                      "color": "#43c3d3",
                      "id": "Investor Equity",
                    },
                    Object {
                      "color": "#8ac13f",
                      "id": "Sponsor Co-Investment",
                    },
                    Object {
                      "color": "#1ea0c0",
                      "id": "Senior Debt",
                    },
                    Object {
                      "parent": "Investor Equity",
                      "value": 31.9,
                    },
                    Object {
                      "parent": "Sponsor Co-Investment",
                      "value": 3.5,
                    },
                    Object {
                      "parent": "Senior Debt",
                      "value": 53,
                    },
                  ],
                  "levels": Array [
                    Object {
                      "dataLabels": Object {
                        "align": "left",
                        "enabled": true,
                        "style": Object {
                          "fontSize": "15px",
                          "fontWeight": "bold",
                        },
                        "verticalAlign": "top",
                      },
                      "layoutAlgorithm": "sliceAndDice",
                      "level": 1,
                    },
                  ],
                  "type": "treemap",
                },
              ],
            }
          }
        />
      </div>
    </div>
    <div
      className="CapitalStack__content"
    >
      <div
        className="CapitalStack__contentItem"
        key="content-Investor Equity"
      >
        <div
          className="CapitalStack__titleWrapper"
        >
          <span
            className="CapitalStack__label"
            style={
              Object {
                "backgroundColor": "#43c3d3",
              }
            }
          />
          Investor Equity
        </div>
        <div
          className="CapitalStack__amountContent"
        >
          31.9%
          <span
            className="CapitalStack__amount"
          >
            $3,190
          </span>
        </div>
      </div>
      <div
        className="CapitalStack__contentItem"
        key="content-Sponsor Co-Investment"
      >
        <div
          className="CapitalStack__titleWrapper"
        >
          <span
            className="CapitalStack__label"
            style={
              Object {
                "backgroundColor": "#8ac13f",
              }
            }
          />
          Sponsor Co-Investment
        </div>
        <div
          className="CapitalStack__amountContent"
        >
          3.5%
          <span
            className="CapitalStack__amount"
          >
            $350
          </span>
        </div>
      </div>
      <div
        className="CapitalStack__contentItem"
        key="content-Senior Debt"
      >
        <div
          className="CapitalStack__titleWrapper"
        >
          <span
            className="CapitalStack__label"
            style={
              Object {
                "backgroundColor": "#1ea0c0",
              }
            }
          />
          Senior Debt
        </div>
        <div
          className="CapitalStack__amountContent"
        >
          53%
          <span
            className="CapitalStack__amount"
          >
            $5,300
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="CapitalStack__footer"
  >
    <div
      className="CapitalStack__footerContent"
    >
      <span>
        Total:
      </span>
      $10,000
      <span>
        (100%)
      </span>
    </div>
  </div>
</div>
`;
